rfcs/text/0003-make-provides-async.md

```diff
@@ -0,0 +1,160 @@
+
+# Basic example
+
+```js
+import App, {createToken} from 'fusion-core';
+
+export default async function start(options: any = {}) {
+  const root = options.root || DefaultRoot;
+  const app = new App(root, options.render);
+
+  const SomeToken = createToken('someToken');
+  app.register(
+    SomeToken,
+    createPlugin({
+      provides: deps => Promise.resolve(42),
+    })
+  );
+  app.register(
+    createPlugin({
+      deps: {someDep: SomeToken},
+      provides: ({someDep}) => {
+        /* Prior to the implementation of this RFC, the following line will
+         * fail as `someDep` is of type `Promise<number>` instead of
+         * `number`.
+         */
+        (someDep: number);
+        return someDep + 5; // 47 instead of "[object Promise]5"
+      },
+    })
+  );
+
+  return app;
+};
+```
+
+# Motivation
+
+This proposal is motivated by situations where a plugin's service requires some initialization that occurs asynchronously (e.g. I/O, requests to back-end services, etc).  Today, there exists workarounds but they require additional boilerplate or are fairly clunky to use, leading to a degraded developer experience.
+
+A few potential workarounds:
+- a service may be wrapped in a `Promise<TService>` that the consumer must resolve prior to usage.
+- a service may expose an asynchronous initialization method that must be resolved prior to usage.  For example, in the `middleware` as part of processing a request.
+- use blocking calls only and avoid the advantages of asynchronous execution.
+
+## Use Cases
+
+There are a number of real life use cases where the app reaches beyond the boundaries of its code.  A few of these use cases are illustrated below.
+
+#### Example #1
+
+A plugin that must communicate with a back-end service in order to initialize its service:
+
+```js
+// old way
+app.register(createPlugin({
+  provides: () => {
+    let isInitialized = false;
+    return {
+      doSomething: () => {
+        if(!isInitialized) throw new Error("Not initialized!");
+        /* some business logic */
+      },
+      initializeAsync: async () => {
+        /* some async call to back-end service */
+        isInitialized = true;
+      }
+    };
+  }
+}));
+```
+
+In the above case, a consuming plugin that directly depends upon the above plugin's service cannot utilize the exported service inside of their service until it has been initialized.  The consuming plugin in this case requires an asynchronous initialization step that initializes the parent service.
+
+```js
+// new way
+app.register(createPlugin({
+  provides: async () => {
+    /* some async call to back-end service */
+    /* e.g. await someInitializationCall(); */
+    return {
+      doSomething: () => {
+        /* some business logic */
+      }
+    };
+  }
+}));
+```
+
+A consuming plugin that depends upon the above plugin will be able to consume the exported service and use it synchronously.
```

Comment:
So if I'm understanding correctly, this handles the `await` internally, so doesn't change the signature for using plugins, just makes it easier to author them?

Reply:
Correct!  The signature for plugins would be unchanged.  It would, however, alter the behavior of an asynchronous service injected as a dependency.  In that case, the promise would be resolved for any consumers (versus today, where it would not).